lms/djangoapps/instructor/lti_grader.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python
+"""
+Send grades to an OpenEdX LTI component
+"""
+import csv
+
+from instructor import lti_connection
+from instructor.lti_connection import LTIConnectionError
+
+
+class LTIGrader(object):
+    """
+    Updates grades for a given LTI component
+    """
+
+    def __init__(self, course_id, url_base, lti_key, lti_secret):
+        self.url_base = url_base
+        self.course_id = course_id
+        self.key = lti_key
+        self.secret = lti_secret
+        self.output = {'success': '', 'error': ''}
+
+    def _get_first_anon_id(self, grade_data):
+        """
+        Returns the first anonymized student id for in the given uploaded csv data file
+        :param grade_data: an InMemoryUploadedFile with lti grade data
+        :return: anon_id
+        """
+        reader = self._unicode_csv_reader(grade_data)
+        try:
+            reader.next()
+            first_row = reader.next()
+            anon_id = first_row[1]
+        except StopIteration:
+            self.output['error'] += '\nThe CSV file contains no useful data!'
+            anon_id = None
+        return anon_id
+
+    def update_grades(self, grade_data):
+        """
+        Updates grades for the LTI component with the grade data provided
+        :param grade_data: an InMemoryUploadedFile with lti grade data
+        :return: output, a dictionary with two keys: 'success' and 'error'
+        """
+        grade_data = grade_data.read().splitlines()
+
+        first_anon_id = self._get_first_anon_id(grade_data)
+        if first_anon_id is None:
+            return self.output
+
+        test_url = self.url_base + first_anon_id
+        try:
+            lti_connection.validate_lti_passport(self.key, self.secret, test_url)
+        except LTIConnectionError as error:
+            self.output['error'] += '\n' + unicode(error)
+            return self.output
+
+        for grade_row in self._generate_valid_grading_rows(grade_data):
+            try:
+                self.output['success'] += '\n' + lti_connection.post_grade(self.url_base, self.key, self.secret, grade_row)
+            except LTIConnectionError as error:
+                self.output['error'] += '\n' + unicode(error)
+        return self.output
+
+    def _generate_valid_grading_rows(self, data):
+        """
+        Yield valid grading rows from a file
+        """
+        reader = self._unicode_csv_reader(data)
+        row_num = 0
+        for row in reader:
+            row_num += 1
+
+            if row_num == 1:
+                continue
+
+            if len(row) < 5:
+                self.output['error'] += "\nBAD ROW: grading_csv row {row_number} ({row}) doesn't have enough info".format(
+                    row_number=row_num,
+                    row=row,
+                )
+                continue
+            try:
+                if len(row) == 5:
```

Comment:
Since we also have special case logic elsewhere for when the `len(row) != 6`, should we just normalize the data structure here and and enforce that the `grade_row` tuple such that it always contains 6 items (using some sane default)?